cdap-master/src/main/java/io/cdap/cdap/master/environment/k8s/AuthenticationServiceMain.java

```diff
@@ -0,0 +1,152 @@
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.master.environment.k8s;
+
+import com.google.common.base.Throwables;
+import com.google.common.util.concurrent.AbstractService;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.Service;
+import com.google.inject.Injector;
+import com.google.inject.Module;
+import io.cdap.cdap.common.ServiceBindException;
+import io.cdap.cdap.common.conf.CConfiguration;
+import io.cdap.cdap.common.conf.Constants;
+import io.cdap.cdap.common.guice.ZKClientModule;
+import io.cdap.cdap.common.logging.LoggingContext;
+import io.cdap.cdap.common.logging.ServiceLoggingContext;
+import io.cdap.cdap.master.spi.environment.MasterEnvironment;
+import io.cdap.cdap.master.spi.environment.MasterEnvironmentContext;
+import io.cdap.cdap.messaging.MessagingService;
+import io.cdap.cdap.messaging.guice.MessagingClientModule;
+import io.cdap.cdap.proto.id.NamespaceId;
+import io.cdap.cdap.security.guice.SecurityModules;
+import io.cdap.cdap.security.impersonation.SecurityUtil;
+import io.cdap.cdap.security.server.ExternalAuthenticationServer;
+import org.apache.twill.internal.Services;
+import org.apache.twill.zookeeper.ZKClientService;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * The main class responsible for Authentication  .
+ */
+public class AuthenticationServiceMain extends AbstractServiceMain<EnvironmentOptions> {
+
+  private static final Logger LOG = LoggerFactory.getLogger(AuthenticationServiceMain.class);
+
+  public static void main(String[] args) throws Exception {
+    main(AuthenticationServiceMain.class, args);
+  }
+
+  @Override
+  protected List<Module> getServiceModules(
+      MasterEnvironment masterEnv, EnvironmentOptions options, CConfiguration cConf) {
+    return Arrays.asList(
+        getDataFabricModule(),
+        new ZKClientModule(),
+        new SecurityModules().getDistributedModules(),
+        new MessagingClientModule());
+  }
+
+  @Override
+  protected void addServices(
+      Injector injector, List<? super Service> services, List<? super AutoCloseable> closeableResources,
+      MasterEnvironment masterEnv, MasterEnvironmentContext masterEnvContext, EnvironmentOptions options) {
+
+    MessagingService messagingService = injector.getInstance(MessagingService.class);
+    if (messagingService instanceof Service) {
+      services.add((Service) messagingService);
+    }
+
+    CConfiguration configuration = injector.getInstance(CConfiguration.class);
+
+    if (configuration.getBoolean(Constants.Security.ENABLED)) {
+
+      services.add(new AbstractService() {
+
+        @Override
+        protected void doStart() {
+
+          ZKClientService zkClientService = injector.getInstance(ZKClientService.class);
```

Comment:
Ideally in k8s environment, we should use ZK for the key store, but instead using k8s secrets. Please open a follow up JIRA for that change. We'll need to update the `MasterEnvironment` to provide an abstraction for secure store.

Reply:
Created - https://cdap.atlassian.net/browse/CDAP-17618